tests/application_development/code_relocation/testcase.yaml

```diff
@@ -6,6 +6,9 @@ tests:
     arch_allow: arm
     extra_configs:
       - CONFIG_RELOCATE_TO_ITCM=y
+      - CONFIG_NULL_POINTER_EXCEPTION_DETECTION_NONE=y
+    platform_exclude:
+      - nucleo_h7a3zi_q
```

Comment:
I rather think that the issue is in the `filter/arch_allow` combination.
Looking to `tests/application_development/code_relocation/CMakeLists.txt` I wonder how the following twister filter is supposed to produce meaningful results on platforms other than NXP.
`zephyr,itcm` is not available only on NXP platforms (nucleo_h7a3zi_q is a good example), yet test will be run on any platform supporting `zephyr,itcm`.
`platform_exclude` is a one bullet gun. It will have to be updated each time wrongly composed `filter/arch_allow` fails to provide correct filtering

Reply:
I port this testcases due to an issue found when code relocate in pre-kernel memcpy usage, so it is limited to NXP platforms, but following enhances seems try to introduce more platforms. But I think this is not related to this PR

Reply:
Problem is I don't understand how this is supposed to work on platforms that are not dealt with in CMakeLists.txt. IMO filter should be much more restrictive.

Reply:
I agree that filters should be more restrictive if this test is meant only to some NXP platforms, but I'd suggest @hakehuang (as initial author of this test) address this in a separate PR to address #60167 and for the many other SoCs this test may not be building. If there's a limited subset of NXP boards this test is addressed, then a `allow_platforms` filter seems more reasonable here.

Reply:
> platform_exclude is a one bullet gun. It will have to be updated each time wrongly composed filter/arch_allow fails to provide correct filtering

Right, the addition of `platform_exclude` here is not meant to be a fix, just a workaround to unblock 6 other PR's affected by this test failure. I'm not familiar with the ST boards so I cannot propose a fix for them and logged it separately in #60167. As I commented there, this is likely to happen in many other Arm boards as well.

Reply:
I think the solution is to restrict to NXP boards for now

Reply:
then we need add platform_allow

Reply:
@hakehuang `frdm_k64f` is not selecting any `zephyr,itcm`. This test cannot link for that board. Can you please double check this is correct?

Reply:
@manuargue , Thanks a lot for this checking, it was a mistaken here please do below changes, and I test is on frdm_k64f whcih works
```
  tests.application_development.code_relocation_kinetis:
    filter: CONFIG_CPU_HAS_NXP_MPU and CONFIG_MINIMAL_LIBC
    arch_allow: arm
    extra_configs:
      - CONFIG_MPU_ALLOW_FLASH_WRITE=y
    platform_allow:
      - frdm_k64f
```

```
*** Booting Zephyr OS build zephyr-v3.4.0-975-gd8b2a7263ba8 ***
Running TESTSUITE code_relocation
===================================================================
START - test_function_genex_relocate_1
Address of var_file4_sram2_data 0x200193a4
Address of var_file4_sram2_bss 0x200193e0

 PASS - test_function_genex_relocate_1 in 0.008 seconds
===================================================================
START - test_function_genex_relocate_2
Address of var_file5_sram2_data 0x200193a8
Address of var_file5_sram2_bss 0x200193f8

 PASS - test_function_genex_relocate_2 in 0.008 seconds
===================================================================
START - test_function_in_split_multiple
Address of var_file3_sram_data 0x20000580
Address of var_file3_sram2_bss 0x200193c8

 PASS - test_function_in_split_multiple in 0.008 seconds
===================================================================
START - test_function_in_sram2
Address of var_sram2_data 0x200193a0
Address of k_sem_give 0x200181c1
Address of var_sram2_rodata 0x20018b40
Address of var_sram2_bss 0x200193b0

Hello World! frdm_k64f
Address of memcpy 0x6303

Address of var_lib1_sram2_data 0x200193ac
Address of var_lib1_sram2_bss 0x20019410
Address of relocated_lib_helper 0x200185fd

Relocated helper function running on frdm_k64f

Address of custom_section, func placed using attributes 0x20001b85
Address of custom_section data placed using attributes 0x20001b80

 PASS - test_function_in_sram2 in 0.046 seconds
===================================================================
TESTSUITE code_relocation succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [code_relocation]: pass = 4, fail = 0, skip = 0, total = 4 duration = 0.070 seconds
 - PASS - [code_relocation.test_function_genex_relocate_1] duration = 0.008 seconds
 - PASS - [code_relocation.test_function_genex_relocate_2] duration = 0.008 seconds
 - PASS - [code_relocation.test_function_in_split_multiple] duration = 0.008 seconds
 - PASS - [code_relocation.test_function_in_sram2] duration = 0.046 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL

```